{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pylab\n",
    "import math\n",
    "import pandas as pd\n",
    "import glob\n",
    "from collections import defaultdict\n",
    "from sourmash.lca import lca_utils\n",
    "from sourmash.lca.lca_utils import LineagePair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "#sample_id='SRR5950647'\n",
    "#outdir = 'outputs.test'\n",
    "sample_id='SRR606249'\n",
    "outdir = 'outputs.paper'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown as md\n",
    "from IPython.display import display\n",
    "md(f\"# genome-grist taxonomic summary for metagenome `{sample_id}`\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load gather CSV + taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load gather CSV\n",
    "tax_df = pd.read_csv(f'../../{outdir}/gathertax/{sample_id}.x.genbank.gather.with-lineages.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tax_ranks = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "\n",
    "# \"fix\" lineage column,\n",
    "def split_lineages(x):\n",
    "#    print((x, type(x)))\n",
    "    if isinstance(x, float) and math.isnan(x):\n",
    "        return ()\n",
    "\n",
    "    ret = []\n",
    "    for (rank, name) in zip(tax_ranks, x.split(';')):\n",
    "        ret.append(LineagePair(rank, name))\n",
    "    return tuple(ret)\n",
    "    \n",
    "tax_df['lineage'] = tax_df['lineage'].apply(split_lineages)\n",
    "\n",
    "# and break it up into tuples\n",
    "\n",
    "def grab_tax(x, idx):\n",
    "    if x:\n",
    "        return x[idx].name\n",
    "    return ''\n",
    "\n",
    "for idx, rank in enumerate(tax_ranks):\n",
    "    tax_df[rank] = tax_df['lineage'].apply(grab_tax, args=(idx,))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create function that aggregates various pieces of information by rank & builds a new df\n",
    "def aggregate_by_rank(df, rank):\n",
    "    # first, build all the counts\n",
    "    unique_intersect_by_rank = defaultdict(int)\n",
    "    f_intersect_by_rank = defaultdict(int)\n",
    "    best_hashes_by_rank = {}\n",
    "\n",
    "    def sum_unique(row):\n",
    "        lin = row.lineage\n",
    "        if pd.isnull(lin):\n",
    "            lin = []\n",
    "        unique_hashes = row.unique_intersect_bp\n",
    "        f_hashes = row.f_unique_to_query\n",
    "\n",
    "        for rank in ('superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'):\n",
    "            poplin = lca_utils.pop_to_rank(lin, rank)\n",
    "            unique_intersect_by_rank[poplin] += unique_hashes\n",
    "            f_intersect_by_rank[poplin] += f_hashes\n",
    "            \n",
    "            best_hashes = best_hashes_by_rank.get(poplin, 0)\n",
    "            if best_hashes < unique_hashes:\n",
    "                best_hashes_by_rank[poplin] = unique_hashes\n",
    "\n",
    "    df.apply(sum_unique, axis=1)\n",
    "\n",
    "    # now do the sorting etc.\n",
    "    rank_counts = []\n",
    "    for lin, bp in sorted(unique_intersect_by_rank.items(), key=lambda x: -x[1]):\n",
    "        if lin and lin[-1].rank == rank:\n",
    "            \n",
    "            f_unique = f_intersect_by_rank[lin]\n",
    "            if rank == 'species':\n",
    "                name = lin[-1].name\n",
    "                name = f'{name} ({lin[0].name})'\n",
    "            elif rank == 'genus':\n",
    "                name = lin[-1].name\n",
    "                name = f'{name}'\n",
    "            else:\n",
    "                name = lca_utils.display_lineage(lin)\n",
    "\n",
    "            rank_counts.append(dict(name=name, hashes=bp, f_unique=f_unique,\n",
    "                                    best_hashes=best_hashes_by_rank[lin],\n",
    "                                    lineage=lin))\n",
    "            \n",
    "    name = \"unclassified\"\n",
    "    hashes = 0 # CTB fixme\n",
    "    f_unique = 1 - sum(v for (k, v) in f_intersect_by_rank.items() if k and k[-1].rank == rank )\n",
    "    best_hashes = 0 # CTB\n",
    "    lineage = ()\n",
    "    rank_counts.append(dict(name=name, hashes=bp, f_unique=f_unique,\n",
    "                            best_hashes=best_hashes_by_rank[lin],\n",
    "                            lineage=lin))    \n",
    "\n",
    "    return pd.DataFrame(rank_counts)\n",
    "\n",
    "species_df = aggregate_by_rank(tax_df, 'species')\n",
    "genus_df = aggregate_by_rank(tax_df, 'genus')\n",
    "family_df = aggregate_by_rank(tax_df, 'family')\n",
    "order_df = aggregate_by_rank(tax_df, 'order')\n",
    "class_df = aggregate_by_rank(tax_df, 'class')\n",
    "phylum_df = aggregate_by_rank(tax_df, 'phylum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'species'\n",
    "plot_df = species_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'genus'\n",
    "plot_df = genus_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'family'\n",
    "plot_df = family_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'order'\n",
    "plot_df = order_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'class'\n",
    "plot_df = class_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank = 'phylum'\n",
    "plot_df = phylum_df\n",
    "\n",
    "pylab.figure(num=None, figsize=(10, 10))\n",
    "\n",
    "plot_df = plot_df[:40]\n",
    "#pylab.plot(plot_df.hashes, plot_df.iloc[::-1].index, 'bo', label='hashes')\n",
    "pylab.plot(plot_df.f_unique * 100, plot_df.iloc[::-1].index, 'gx', label='percent of metagenome')\n",
    "\n",
    "positions = list(plot_df.index)\n",
    "labels = list(reversed(plot_df.name))\n",
    "pylab.yticks(positions, labels, fontsize='small')\n",
    "\n",
    "pylab.xlabel('bp')\n",
    "pylab.axis(xmin=0)\n",
    "pylab.legend(loc='lower right')\n",
    "pylab.title(f'{sample_id}: metagenome classification, by {rank}')\n",
    "pylab.tight_layout()\n",
    "\n",
    "#figname = f'/tmp/gathergram-{rank}-{sample_id}.png'\n",
    "#print(f'saving to {figname}')\n",
    "#pylab.savefig(figname)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (myenv)",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
