<!DOCTYPE html>
<html lang="en" data-bs-theme="light">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Configuring - genome-grist docs</title>
        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/fontawesome.min.css" rel="stylesheet">
        <link href="../css/brands.min.css" rel="stylesheet">
        <link href="../css/solid.min.css" rel="stylesheet">
        <link href="../css/v4-font-face.min.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link id="hljs-light" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" >
        <link id="hljs-dark" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github-dark.min.css" disabled>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script>hljs.highlightAll();</script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="..">genome-grist docs</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-bs-toggle="collapse" data-bs-target="#navbar-collapse" aria-controls="navbar-collapse" aria-expanded="false" aria-label="Toggle navigation">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="nav-item">
                                <a href=".." class="nav-link">Home</a>
                            </li>
                            <li class="nav-item">
                                <a href="../quickstart/" class="nav-link">Quickstart</a>
                            </li>
                            <li class="nav-item">
                                <a href="./" class="nav-link active" aria-current="page">Configuring</a>
                            </li>
                            <li class="nav-item">
                                <a href="../output-guide/" class="nav-link">Output files</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ms-md-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-bs-toggle="modal" data-bs-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../quickstart/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" href="../output-guide/" class="nav-link">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-bs-toggle="collapse" data-bs-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-body-tertiary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-bs-level="1"><a href="#configuring-a-genome-grist-project" class="nav-link">Configuring a genome-grist project</a>
              <ul class="nav flex-column">
            <li class="nav-item" data-bs-level="2"><a href="#overview" class="nav-link">Overview</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#analyzing-your-own-metagenomes" class="nav-link">Analyzing your own metagenomes</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#using-genbank-genomes" class="nav-link">Using Genbank genomes</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#preparing-information-on-local-genomes" class="nav-link">Preparing information on local genomes</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#an-example-for-you-to-try-the-podar-ref-database" class="nav-link">An example for you to try: the podar-ref database</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#reference-the-complete-set-of-config-file-options" class="nav-link">Reference: The complete set of config file options</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#more-advanced-genome-grist-usage" class="nav-link">More advanced genome-grist usage</a>
              <ul class="nav flex-column">
              </ul>
            </li>
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="configuring-a-genome-grist-project">Configuring a genome-grist project</h1>
<!-- CTB: this is doc/configuring.md in dib-lab/genome-grist -->

<p><a href="https://hackmd.io/p7QfD_SsQg6sElDbrzpcsA"><img alt="hackmd-github-sync-badge" src="https://hackmd.io/p7QfD_SsQg6sElDbrzpcsA/badge" /></a></p>
<p>[toc]</p>
<h2 id="overview">Overview</h2>
<p>genome-grist does the following:</p>
<ul>
<li>downloads metagenome data from the SRA, if requested;</li>
<li>pre-processes and trims each metagenome;</li>
<li>runs <code>sourmash gather</code> on each metagenome, using one or more sourmash databases, to find a <a href="https://www.biorxiv.org/content/10.1101/2022.01.11.475838v1">minimum metagenome cover</a>;</li>
<li>retrieves the full genomes for any matches and executes a variety of mapping-based analyses;</li>
<li>incorporates taxonomy information for the genomes into the taxonomy summary report (if taxonomy reporting is requested).</li>
</ul>
<p>Much of the configuration for genome-grist is about where to find more information about matching genomes - specifically, the genomes, their display names, and their taxonomy.</p>
<p>For Genbank genomes, this is easy and genome-grist does it automatically! But if you're providing your own genomes and taxonomy information, it's a bit trickier.</p>
<h2 id="analyzing-your-own-metagenomes">Analyzing your own metagenomes</h2>
<p>You can provide a list of SRA run accessions in the <code>samples:</code> list in
the config file, and genome-grist will automatically download, interleave,
and trim them for you.</p>
<p>If you want to run genome-grist on your own metagenomes, you need to
provide one FASTQ file per sample in the <code>trim/</code> subdirectory of the
output directory; for example, for the output directory
<code>outputs.private</code> and the sample named <code>podar</code>, you would need to
create <code>outputs.private/trim/podar.trim.fq.gz</code>.  This should be an
interleaved file of Illumina reads, as generated by (for example)
<code>seqtk mergepe</code>.  The file must start with the sample name and end in
<code>trim.fq.gz</code>.</p>
<p>Providing correctly-named files will shortcut automatic SRA downloads for
these files, and genome-grist will download any remaining samples.</p>
<p>If you want to prevent automatic downloading from the SRA completely,
you can set the parameter <code>prevent_sra_download: true</code> in the config file.
This is a good parameter to set if you are only analyzing your own
prepared data!</p>
<h2 id="using-genbank-genomes">Using Genbank genomes</h2>
<p>For Genbank genomes, all the necessary information is available already, or automatically determined by genome-grist.</p>
<p>sourmash already provides
<a href="https://sourmash.readthedocs.io/en/latest/databases.html">pre-built databases containing all GTDB genomes (R07 rs207) as well as 1.3m Genbank microbial genomes from 2022</a>.</p>
<p>For genomes available through Genbank (aka with Genbank accessions), genome-grist does the genome retrieval automatically, so you don't need to have them downloaded already.</p>
<p>Taxonomy spreadsheets are available for both GTDB and Genbank at the Databases page above.</p>
<h2 id="preparing-information-on-local-genomes">Preparing information on local genomes</h2>
<p>You can provide your own sourmash databases, your own set of genome files and genome information and your own taxonomy spreadsheet to genome-grist,
too! This can be useful if you have unpublished or private genomes that you want to use with genome-grist, or if you have large subsets of Genbank already downloaded. Luckily this is all reasonably straightforward and we provide tools to help you! Read on!</p>
<p>Note that you can absolutely combine Genbank with your own databases here, or just use your own databases. (If there are overlapping identifiers, the local genomes are chosen first; you might want to do this if you already have a bunch of the Genbank genomes downloaded already, for example.)</p>
<h3 id="choosing-identifiers-for-your-genomes">Choosing identifiers for your genomes</h3>
<p>You'll need to choose unique identifiers for your genomes. genome-grist requires that your identifier does not have a space, colon (:), or forward slash (/) in it; everything else should be fine.</p>
<h3 id="preparing-your-genome-files">Preparing your genome files</h3>
<p>You'll need one FASTA file per genome (gzip or bz2 compressed is fine). The filename doesn't matter. It's probably easiest if they're all in one directory, although this isn't necessary.</p>
<p>For now, we suggest naming the first sequence in each FASTA file with the genome identifier at the start, space delimited - for example <code>MY_ID_1.1 first_sequence_name is very special</code>. This will allow sourmash to name the resulting signature with the right identifier using <code>--name-from-first</code> (see below).</p>
<h3 id="creating-one-or-more-sourmash-databases">Creating one or more sourmash databases</h3>
<p>You can mix local databases with genbank databases without fear! You'll need to provide one or more sourmash databases for any local collections, and you do this as usual via the config parameter <code>sourmash_databases</code>, which takes a list of paths to sourmash database locations.</p>
<p>To build your own sourmash databases, you'll need sourmash sketches
for each genome. Sketch all your genomes (in <code>fna.gz</code> or <code>.fa.gz</code>
files) with the following command:</p>
<pre><code>sourmash sketch dna -p k=31,scaled=1000 *.gz
</code></pre>
<p>If you've named your genomes so that the first sequence contains the identifier, you can add <code>--name-from-first</code> and then the signatures will be named the right thing for the next step.</p>
<p>If not, you'll need to manually rename of the signatures produced by <code>sourmash sketch</code>. (You can do this with <code>sourmash sig rename</code>, but there's no simple way to do this in bulk.)</p>
<p>Once you have all your genome signatures, you can create a sourmash database with</p>
<pre><code>sourmash index output.sbt.zip *.sig
</code></pre>
<p>and then you can cherish and treasure your sourmash database forever!</p>
<p>If you have lots of genomes (1000 or more?) we suggest using a workflow system to sketch and rename your genomes appropriately; please ask us for some examples over on <a href="https://github.com/dib-lab/sourmash/issues">the sourmash issue tracker</a>.</p>
<p>We chose k=31 above (in the <code>sourmash sketch</code> command) because that matches our default parameters, and we have provided Genbank and GTDB databases for k=31 (as well as k=21 and k=51). But the only real requirement here is that all your databases support the same requested k-mer and scaled sizes.</p>
<h3 id="providing-your-local-genomes-to-genome-grist">Providing your local genomes to genome-grist</h3>
<p>You'll also need to provide your local genome files to genome-grist, along with their "display name", which will be used in reporting.  The information will be provided via the config parameter <code>local_databases_info</code>, which takes a list of paths to info file CSVs.</p>
<p><strong>First</strong>, genome-grist needs the genome sequences in their own individual files, in one or more directories. The files need to be named by their identifiers in the format <code>{ident}_genomic.fna.gz</code>, and must come with an "info file" that contains their identifier, a display name, and the location of the genome file (which <em>must</em> be named as above).</p>
<p>genome-grist has a utility to help set this all up! The script <code>genome_grist.copy_local_genomes</code> will take in a list of FASTA files containing genome(s), read the header of the first sequence to find the identifier for that genome, and then copy it into a directory for you. (see "Step 3", below, for execution instructions for this script). It will also output a provisional info file, which you can edit.</p>
<p>Here's an example of the output of the info file produced by <code>copy_local_genomes</code>:</p>
<pre><code>ident,display_name,genome_filename
CP001472.1,&quot;Acidobacterium capsulatum ATCC 51196, complete genome&quot;,databases/podar-ref.d/CP001472.1_genomic.fna.gz
CP001941.1,&quot;Aciduliprofundum boonei T469, complete genome&quot;,databases/podar-ref.d/CP001941.1_genomic.fna.gz
CP001097.1,&quot;Chlorobium limicola DSM 245, complete genome&quot;,databases/podar-ref.d/CP001097.1_genomic.fna.gz
</code></pre>
<p><strong>Second</strong>, for each genome, genome-grist also needs a separate <code>{ident}.info.csv</code> file, containing just the identifier and the display name. This needs to be in the same directory as the genome itself.</p>
<p>The utility script <code>genome_grist.make_info_file</code> will produce this for you, based on the whole-database info CSV file created above. (See "Step 4", below, for execution instructions for this script.)</p>
<p>Here's an example of the output of <code>make_info_file:</code>; this is the file <code>CP001097.1.info.csv</code>:</p>
<pre><code>ident,display_name
CP001097.1,&quot;Chlorobium limicola DSM 245, complete genome&quot;
</code></pre>
<p>and the final contents of the <code>databases/podar-ref.d/</code> directory include:</p>
<pre><code>CP001097.1.info.csv             CP001472.1_genomic.fna.gz
CP001097.1_genomic.fna.gz       CP001941.1.info.csv
CP001472.1.info.csv             CP001941.1_genomic.fna.gz
</code></pre>
<h3 id="providing-taxonomy-information">Providing taxonomy information</h3>
<p>If you want to enable taxonomic summarization for your local genomes, you'll need a taxonomy file that can be read by the <code>sourmash tax</code> subcommands - see <a href="https://sourmash.readthedocs.io/en/latest/command-line.html">the sourmash command-line docs</a> for details. This file contains at least 8 columns, with the headers <code>ident</code> and <code>superkingdom</code>, <code>phylum</code>,<code>class</code>,<code>order</code>,<code>family</code>,<code>genus</code>,<code>species</code>. You provide this file to genome-grist via the config parameter <code>taxonomies</code>, which takes a list of paths to sourmash taxonomy files.</p>
<h3 id="testing-it-all-out">Testing it all out</h3>
<p>We recommend trying this all out with a fake metagenome that's just two of your local genomes concatenated; you can set this up by making the FASTA file and then putting it in your output directory in the subdirectory <code>trim/{sample}.trim.fq.gz</code>, and configuring genome-grist to run <code>summarize_gather</code> on just that sample.</p>
<p>So, for example, </p>
<ul>
<li>create a file <code>trim/testme.trim.fq.gz</code> containing a bunch of sequences (FASTA or FASTQ format, despite the filename :)</li>
<li>set <code>samples</code> in your config file <code>conf-test.yml</code> to <code>- testme</code></li>
<li>run <code>genome-grist run conf-test.yml summarize_gather</code></li>
</ul>
<p>and if it all works, then your local database configuration is good! (The output report will be in the <code>reports/report-gather-testme.html</code> subdirectory in your output directory.)</p>
<p>You will need to run <code>summarize_tax</code> to test the taxonomy file; the associated output will be in <code>reports/report-taxonomy-testme.html</code>.</p>
<p>If you run into any problems, please <a href="https://github.com/dib-lab/genome-grist/issues">file an issue!</a></p>
<h2 id="an-example-for-you-to-try-the-podar-ref-database">An example for you to try: the <code>podar-ref</code> database</h2>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/23387867/">Comparative metagenomic and rRNA microbial diversity characterization using archaeal and bacterial synthetic communities, Shakya et al., 2014</a> made a lovely mock metagenome containing approximately 65 different strains of microbes.</p>
<p><a href="https://www.biorxiv.org/content/10.1101/155358v3">Evaluating Metagenome Assembly on a Simple Defined Community with Many Strain Variants, Awad et al., 2017</a> used sourmash to analyze this community, and produced an updated list of reference genomes that is available for  download.</p>
<p>While this list of reference genomes is in fact in Genbank, they use non-Genbank identifiers, and so it's a good example data set for "private" genomes.</p>
<p>So! Let's run through setting up these reference genomes as a local, non-Genbank database for genome-grist to use, and then test it out by applying genome-grist to the mock metagenome!</p>
<p>It should take under 10 minutes total to run all the commands.</p>
<p>Note: If you have a developer installation of <code>genome-grist</code>, you can run everything below with <code>make test-private</code> in the root <code>genome-grist/</code> directory.</p>
<h3 id="step-0-install-genome-grist-and-set-up-your-directory">Step 0: Install genome-grist and set up your directory</h3>
<p>Follow the installation instructions for genome-grist and make sure
you're in a conda environment where genome-grist is installed.</p>
<p>You will also need sourmash...</p>
<pre><code>pip install sourmash
</code></pre>
<p>and now you should be good to go!</p>
<h3 id="step-1-download-and-unpack-the-podar-reference-genomes">Step 1: Download and unpack the <code>podar</code> reference genomes</h3>
<p>First, we need to get our hands on the genome sequences themselves.</p>
<p>The genomes from Awad et al., 2017, are available for download <a href="https://osf.io/vbhy5/">from a project on the Open Science Framework</a>. The following commands will download them and unpack them into the directory <code>databases/podar-ref/</code></p>
<pre><code>mkdir -p databases/podar-ref
curl -L https://osf.io/vbhy5/download -o databases/podar-ref.tar.gz
cd databases/podar-ref/ &amp;&amp; tar xzf ../podar-ref.tar.gz
cd ../../
</code></pre>
<h3 id="step-2-build-sketches-and-construct-a-sourmash-database">Step 2: Build sketches and construct a sourmash database</h3>
<p>genome-grist uses sourmash to generate a <em>minimum metagenome cover</em> containing the best matches to the metagenome, so we need to turn the downloaded genomes into a sourmash database.</p>
<p>The following command will sketch all of the <code>.fa</code> files and save the resulting sourmash signatures into <code>databases/podar-ref.zip</code>:</p>
<pre><code class="language-shell">sourmash sketch dna -p k=31,scaled=1000 --name-from-first \
        databases/podar-ref/*.fa -o databases/podar-ref.zip
</code></pre>
<p>note the use of <code>--name-from-first</code>, which names the sketches after the first FASTA header in each file.</p>
<p>If you look at the zip file with <code>sourmash sig describe databases/podar-ref.zip</code>, you'll see that all of the signature names start with their accessions, which is what we want.</p>
<h3 id="step-3-copy-the-genomes-in-to-a-new-location-with-new-names">Step 3: Copy the genomes in to a new location with new names</h3>
<p>Copy the local genomes to a new home that matches genome-grist requirements like so:</p>
<pre><code>python -m genome_grist.copy_local_genomes databases/podar-ref/*.fa -o databases/podar-ref.info.csv -d databases/podar-ref.d
</code></pre>
<p>The subdirectory <code>databases/podar-ref.d/</code> should now contain 64 genome files, named by their identifiers. </p>
<p>There will also be an "information file", <code>databases/podar-ref.info.csv</code>, that contains three columns. These were auto-generated by the script from the FASTA files you gave it. You can edit the <code>display_name</code> column and change it to whatever you want; the other columns need to match with other information so please don't change those!</p>
<p>Note that <code>display_name</code> is just for display purposes; this allows grist to translate identifiers to (for example) a species and strain name to put on generated graphs.</p>
<h3 id="step-4-build-genome-info-files-for-genome-grist">Step 4: Build genome "info files" for genome-grist</h3>
<p>Next, you'll need to create <code>{ident}.info.csv</code> files for each genome. Run:</p>
<pre><code>python -m genome_grist.make_info_file databases/podar-ref.info.csv
</code></pre>
<p>to use the combined info CSV from the previous step to create the necessary info files.</p>
<p>The subdirectory <code>databases/podar-ref.d/</code> should now contain 128 files - 64 genome files, and 64 '.info.csv' files, one for each genome.</p>
<h3 id="step-5-download-the-taxonomy-file">Step 5: Download the taxonomy file</h3>
<p>Last but not least, you'll want a taxonomy file for these genomes, in a format that <code>sourmash taxonomy</code> can use.  For this data set, you can get it <a href="https://osf.io/4yhjw/">from a project on the Open Science Framework</a>.</p>
<p>To download it, run:</p>
<pre><code>curl -L https://osf.io/4yhjw/download -o databases/podar-ref.tax.csv
</code></pre>
<p>This will create a local CSV file with superkingdom, phylum, etc. entries for each of the reference genomes you've downloaded.</p>
<h3 id="step-6-try-it-out-on-a-small-mock-metagenome">Step 6: Try it out on a (small) mock metagenome!</h3>
<p>While you can certainly run this on the entire metagenome from Shakya et al., 2014, that will take a while. So we've prepared a 1m read subset of the data for you to try out! Exciting!</p>
<p>You can download this subsetted metagenome like so:</p>
<pre><code class="language-shell">mkdir -p outputs.private/trim
curl -L https://osf.io/ckbq3/download -o outputs.private/trim/podar.trim.fq.gz
</code></pre>
<p>and then confirm that the config file <code>conf-private.yml</code> has the following content:</p>
<pre><code class="language-yaml">samples:
- podar

outdir: outputs.private/

sourmash_databases:
- databases/podar-ref.zip

local_databases_info:
- databases/podar-ref.info.csv

taxonomies:
- databases/podar-ref.tax.csv
</code></pre>
<p>Now run:</p>
<pre><code>genome-grist run conf-private.yml summarize_gather summarize_mapping -j 4 -p
</code></pre>
<p>and (hopefully) it will all work!!</p>
<p>Assuming there are no errors and everything is green, look at the HTML files in <code>outputs.private/reports/*.html</code>.</p>
<h2 id="reference-the-complete-set-of-config-file-options">Reference: The complete set of config file options</h2>
<p>The options below can be set and/or overridden in a project specific config file that is passed into <code>genome-grist</code>.</p>
<p>Config files can be either <a href="https://en.wikipedia.org/wiki/YAML">YAML</a> or JSON. We suggest YAML since it's nicer to edit.</p>
<p>Every genome-grist installation comes with two config files in the <code>conf/</code> subdirectory of the <code>genome_grist/</code> Python package, <code>defaults.conf</code> and <code>system.conf</code>.  They are read in the order <code>defaults.conf</code>, <code>system.conf</code>, and project-specific config. So, you can ignore the first two and just override options in the project-specific config file. But you can also change the install-wide default parameters in <code>system.conf</code> if you like.</p>
<p>You can use <code>showconf</code> to show the current aggregate config like so: <code>genome-grist run conf.yml showconf</code>.</p>
<h3 id="an-annotated-config-file">An annotated config file</h3>
<pre><code class="language-yaml"># NOTE: all paths are relative to the working directory.

### PROJECT-SPECIFIC PARAMETERS YOU MUST SET FOR EACH PROJECT

# samples: a list of metagenome names. REQUIRED.
# - the sample names cannot contain periods
# - you can use SRA accessions for automatic download, or provide the reads yourself
samples:
- metagenome_one
- metagenome_two

# outdir: a directory where all the output will be placed, e.g. outputs.myproject. REQUIRED.
# this will be created if it doesn't exist.
outdir: some_directory

# metagenome_trim_memory: how much memory (RAM) to use when trimming reads with khmer's trim-low-abund. @CTB
# set to 1e9 for very low diversity samples,
# 10e9 for medium-diversity samples,
# and 50e9 if you're foolishly working with soil :)
# The default is set to 1e9, which is too low for your data.
# WARNING: this much memory _will_ be allocated when running genome-grist!
metagenome_trim_memory: 10e9

### INSTALLATION INFORMATION YOU NEED TO SET AT LEAST ONCE
#
# These must be set after you install genome-grist and download the various databases.

# sourmash_databases: a list of sourmash databases
# you'll need to point this at a local download of 
# databases from e.g. https://sourmash.readthedocs.io/en/latest/databases.html
# cannot be empty!
sourmash_databases:
- /path/to/sourmash-db/database1
- /path/to/sourmash-db/database2

# taxonomies: a list of files to use for taxonomy information. See documentation for `sourmash taxonomy`.
# can be empty list, [].
taxonomies:
- /path/to/taxonomy/files

### INTERMEDIATE CONFIGURATION OPTIONS
#
# These are ways you can fine-tune genome-grist.
# We suggest changing these only once you've successfully run genome-grist a few times!

# local_databases_info: a list of database info files for genomes that are local and/or cannot be downloaded from Genbank.
# can be empty list, [].
# see documentation for more details.
local_databases_info: 
- /path/to/local-sourmash-db/database3.info.csv

# prevent_sra_download: turn off download of metagenomes from SRA by sample ID.&quot;
# DEFAULT: false.
prevent_sra_download: false

# picklist: a --picklist argument to use when searching the sourmash database, to limit which signatures to search.
# see sourmash command line documentation for more details.
# EXAMPLE:
#     picklist: some_sig_list.csv:ident:ident
picklist: &quot;&quot;

# skip_genomes: identifiers to ignore when they show up in gather output.
# This is useful when the sourmash database contains genomes that are no
# longer present in GenBank because they have been deprecated or suppressed.
#
# Note, in such cases you should try to find a new genome to include in
# a local database!
#
# DEFAULT: []
skip_genomes: []

# sourmash_database_threshold_bp: sets the --threshold-bp minimum match
# size of sourmash prefetch and gather. Matches with smaller overlaps
# than this will be excluded from consideration.
# DEFAULT: 1e5
sourmash_database_threshold_bp: 100000

# sourmash_database_ksize: k-mer size to use when searching sourmash databases.
# DEFAULT: 31
sourmash_database_ksize: 31

# sourmash_compute_ksizes: a list of k-mer sizes
# to use when creating sketches for samples. should include the database ksize.
# DEFAULTS: 21, 31, 15
sourmash_compute_ksizes:
- 21
- 31
- 51

# sourmash_scaled: a scaled parameter to use when creating sketches for samples. See sourmash docs for details.
# DEFAULT: 1000
sourmash_scaled: 1000

# sourmash_sigtype: 'DNA' or 'protein' - the type of signature to compute for samples.
# DEFAULT: DNA
sourmash_sigtype: DNA

### SYSTEM SPECIFIC PARAMETERS
#
# These are good defaults for small projects, but you may
# want to change them if you're doing big things on a cluster, or something.

# tempdir: a directory where SRA download temporary files will go, e.g. /tmp
# new subdirs will be created, used, and then removed.
tempdir: some_other_directory

# genbank_cache: where genomes downloaded from genbank will be cached.
# this needs to be writable by people executing genome-grist; if it's system-wide, suggest making a a+rwxt directory.
# DEFAULT ./genbank_cache
genbank_cache: ./genbank_cache

### ADVANCED TECHNICAL PARAMETERS
#
# These probably don't need to be changed unless
# you actually run into problems running genome-grist.

# prefetch_memory: how much memory to allow for
# sourmash prefetch when running genome-grist.
# this memory may not actually be used, depending on sourmash databases used.
# the default is set for the all-Genbank database.
# DEFAULT: 100e9
prefetch_memory: 100e9
</code></pre>
<h2 id="more-advanced-genome-grist-usage">More advanced genome-grist usage</h2>
<h3 id="where-to-insert-your-own-files">Where to insert your own files</h3>
<p>genome-grist is built on top of <a href="https://snakemake.readthedocs.io/en/stable/">the snakemake workflow</a>, which lets you substitute your own files in many places.</p>
<p>For example,</p>
<ul>
<li>you can put your own <code>{sample}_1.fastq.gz</code>, <code>{sample}_2.fastq.gz</code>, and <code>{sample}_unpaired.fastq.gz</code> files in <code>raw/</code> to have genome-grist process reads for you.</li>
<li>you can put your own interleaved reads file in <code>trim/{sample}.trim.fq.gz</code> to run genome-grist on an unpublished or already-preprocessed set of reads;</li>
<li>you can put your own sourmash signature (k=31, scaled=1000) in <code>sigs/{sample}.trim.sig.zip</code> if you want to have it do the database search for you;</li>
</ul>
<p>Please see <a href="https://github.com/dib-lab/genome-grist/blob/latest/genome_grist/conf/Snakefile">the genome-grist Snakefile</a> for all the gory details.</p></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script src="../js/bootstrap.bundle.min.js"></script>
        <script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../js/base.js"></script>
        <script src="../search/main.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
                <p>From here you can search these documents. Enter your search terms below.</p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results" data-no-results-text="No results found"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
